lib/einhorn/command.rb

```diff
@@ -488,6 +492,40 @@ def self.cull
         Einhorn.log_info("Have too many workers at the current version, so killing off #{excess.length} of them.")
         signal_all("USR2", excess)
       end
+
+      # Cleanup any workers that have been signaled but not shutdown yet.
+      kill_old_upgraded_workers if Einhorn::State.signal_timeout
+    end
+
+    def self.kill_old_upgraded_workers
+      now = Time.now
+      children = Einhorn::State.children.select do |_,c|
+        # Only interested in old, upgraded workers that have been signaled.
+        next unless c[:signaled] && c[:signaled].length > 0
+        next unless c[:signaled].include?('USR2')
+
+        # Now filter out those children that have not reached signal_timeout yet.
+        next unless c[:last_signaled_at]
+        expires_at = c[:last_signaled_at] + Einhorn::State.signal_timeout
+        next unless now >= expires_at
+
+        true
+      end
+
+      Einhorn.log_info("#{children.size} old upgraded workers found.") if children.size > 0
+
+      children.each do |pid, child|
+        Einhorn.log_info("Child #{pid.inspect} was signaled #{(child[:last_signaled_at] - now).abs.to_i}s ago. Sending SIGKILL as it is still active after #{Einhorn::State.signal_timeout}s timeout.", :upgrade)
+        begin
+          Process.kill('KILL', pid)
+        rescue Errno::ESRCH
+          Einhorn.log_debug("Attempted to SIGKILL child #{pid.inspect} but the process does not exist.")
+          mourn(pid)
```

Comment:
I belieeeeve we should do here some of what `reap` does, or defer to the later stages of the event loop, where `reap` will reap the exited child processes (otherwise their dead PID will show up in later runs of the event loop and cause `mourn` to get called again).

Reply:
Looking at this a bit more, I am not sure `mourn` should be called here at all (The timer in `signal_all` doesn't do it either).